tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,581 @@
+        update.
+      ma_intervals: number of steps between two "average op", which specifies
+        how frequent a model synchronization is performed.
+      total_num_replicas: Total number of tasks/workers/replicas, could be
+        different from replicas_to_aggregate.
+        total_num_replicas > replicas_to_aggregate: it is backup_replicas +
+        replicas_to_aggregate.
+      block_momentum_rate: It brings in the historical blockwise gradients.
+        The default value is 0.0. When using default value, the naive
+        ModelAverage method is applied, and the original learning rate of
+        local optimizer should be multiply by num_of_workers.
+        If using BMUF algorithm, the block momentum_rate is usually set
+        according to the number of workers: block_momentum_rate =
+        1.0 - 1.0/num_of_workers, the learning rate of local optimizer can be
+        unchanged.
+        For details, see Ref: K. Chen and Q. Huo, "Scalable training of deep
+        learning machines by incremental block training with intra-block
+        parallel optimization and blockwise model-update filtering," in
+        Proceedings of ICASSP, 2016.
+      use_Nesterov: means the Nesterov-style momentum update is applied on the
+        block level. The default value is true. This can accelerate training
+        with non-zero block_momentum_rate.
+      block_learning_rate: block_learning_rate is always 1.0 or slightly higher
+        than 1.0
+    """
+    if total_num_replicas is None:
+      total_num_replicas = replicas_to_aggregate
+    logging.info(
+        "ModelAverageV1: replicas_to_aggregate=%s; total_num_replicas=%s",
+        replicas_to_aggregate, total_num_replicas)
+    self._replicas_to_aggregate = replicas_to_aggregate
+    self._block_momentum_rate = block_momentum_rate
+    self._block_learning_rate = block_learning_rate
+    self._ma_intervals = ma_intervals
+    self._use_Nesterov = use_Nesterov
+    self._gradients_applied = False
+    self._total_num_replicas = total_num_replicas
+    self._tokens_per_step = max(total_num_replicas, replicas_to_aggregate)
+    self._global_step = None
+    self._sync_token_queue = None
+
+    # The synchronization op will be executed in a queue runner which should
+    # only be executed by one of the replicas (usually the chief).
+    self._chief_queue_runner = None
+
+    # Remember which accumulator is on which device to set the initial step in
+    # the accumulator to be global step. This list contains list of the
+    # following format: (accumulator, device).
+    self._accumulator_list = []
+    # ma_run_hook should be called before SyncReplicasHook
+    self._ma_run_hook = False
+    # name: string. Name of the global variables and related operation on ps.
+    self._name = 'modelAverage'
+
+    self.add_variables_for_betweenGraph()
+
+  def add_variables_for_betweenGraph(self):
+    # change all variables to local variables
```

Comment:
I apologize for the all the superficial nit-picky feedback, but we have rather strict style requirements. Comments should be capitalized and punctuated properly.

And the function name should not contain camel-case.